prettyprinter/test/Testsuite/Main.hs

```diff
@@ -86,20 +89,59 @@ tests = testGroup "Tests"
 fusionDoesNotChangeRendering :: FusionDepth -> Property
 fusionDoesNotChangeRendering depth
   = forAll document (\doc ->
-        let rendered = render doc
+    forAll (layouter :: Gen (LayoutOptions -> Doc Int -> SimpleDocStream Int)) (\layouter_ ->
+    forAll arbitrary (\layoutOptions ->
+        let render = renderStrict . layouter_ layoutOptions
+            rendered = render doc
             renderedFused = render (fuse depth doc)
         in counterexample (mkCounterexample rendered renderedFused)
-                          (render doc == render (fuse depth doc)) )
+                          (render doc == render (fuse depth doc)) )))
   where
-    render = renderStrict . layoutPretty defaultLayoutOptions
     mkCounterexample rendered renderedFused
-      = (T.unpack . render . vsep)
+      = (T.unpack . renderStrict . layoutPretty defaultLayoutOptions . vsep)
             [ "Unfused and fused documents render differently!"
             , "Unfused:"
             , indent 4 (pretty rendered)
             , "Fused:"
             , indent 4 (pretty renderedFused) ]
 
+layouter :: CoArbitrary ann => Gen (LayoutOptions -> Doc ann -> SimpleDocStream ann)
+layouter = oneof
+    [ pure layoutPretty
+    , pure layoutSmart
+    , pure (const layoutCompact)
+    -- , Internal.layoutWadlerLeijen <$> arbitrary -- too inconsistent for fusionDoesNotChangeRendering
+    ]
+
+instance Arbitrary LayoutOptions where
+    arbitrary = LayoutOptions <$> oneof
+        [ AvailablePerLine <$> arbitrary <*> arbitrary
+        -- , pure Unbounded -- https://github.com/quchen/prettyprinter/issues/91
```

Comment:
If I uncomment this, the fusion tests trigger #91:

```
    Shallow fusion does not change rendering: FAIL (0.02s)
      *** Failed! (after 56 tests):
      Exception:
        »SFail« must not appear in a rendered »SimpleDocStream«. This is a bug in the layout algorithm! Please report this as a bug
        CallStack (from HasCallStack):
          error, called at src/Data/Text/Prettyprint/Doc/Render/Util/Panic.hs:13:21 in prettyprinter-1.5.1-inplace:Data.Text.Prettyprint.Doc.Render.Util.Panic
      oakland <(commence   inverse)>buttonaccrue tapewormtrackercrowfoot
      spheroid ( printer ` ""sailboatbackfieldendorse transit gazelletacticsdrumbeat  dogsled  dropper puppy dashboard  
       skydive miser eyeglass apple guidancetonic
      oakland southward )
      woodlark
      minnow
      slowdown
      <function>
      LayoutOptions {layoutPageWidth = Unbounded}
      Exception thrown while showing test case:
        »SFail« must not appear in a rendered »SimpleDocStream«. This is a bug in the layout algorithm! Please report this as a bug
        CallStack (from HasCallStack):
          error, called at src/Data/Text/Prettyprint/Doc/Render/Util/Panic.hs:13:21 in prettyprinter-1.5.1-inplace:Data.Text.Prettyprint.Doc.Render.Util.Panic
      
      Use --quickcheck-replay=857723 to reproduce.
```

Two things could IMHO help analyzing these failures:

1. Shrinking: I wonder whether it might be easiest to just use `hedgehog` instead of `QuickCheck`, since that has built-in shrinking. What do you think about this @quchen?

2. A diagnostic representation of the `Doc` (#90).

Reply:
The `arabitrary` should be explicit `range`s, otherwise you get ribbons of 1e-5 and 2 for example.

Reply:
> The `arabitrary` should be explicit `range`s, otherwise you get ribbons of 1e-5 and 2 for example.

Why would that be a problem? I don't really understand the ribbon width logic so far, but I know that `layoutWadlerLeijen` clamps the result to a valid range:

https://github.com/quchen/prettyprinter/blob/7da3b1d32f8a74efb3bcf1b2d062f8f24a64c918/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1810-L1814

I usually try to not to constrain the generated values unless they cause problems – "nonsense" inputs are often helpful at finding bugs in edge cases.